### PR TITLE
Sweep initial final states during training

### LINKED ISSUE #7: sweep x0 and xf during training
I'd like for `x0` and `xf` to change during the training process. Intuition says not much will change in long time horizons, because looking at the DARE calculation there is no usage of `x0` and `xf`. So I expect `K` to converge to the same result, regardless of `x0` and `xf`. In fact, `K` might converge faster if there are more variations. 

I see a problem with the way we are calculating the trajectory costs (and solving this may allow a time varying `xf` as well)/  The way we have defined the TrajectoryCost function, it's dependent on `xf`, which of course is good, but during training and updating it should be updating `xf`. So we need a method to update and/or redefine the `state_costs`.
https://github.com/jsalfity/sensitivity_games/blob/main/sensitivity_games/experiments/point_mass_regulation_train.py#L49-L53 and https://github.com/jsalfity/sensitivity_games/blob/main/sensitivity_games/trajectory_cost.py

Then another problem is updating `xf` in the controller. Again, we'd need a function to update `xf`.
https://github.com/jsalfity/sensitivity_games/blob/main/sensitivity_games/regulator_controller.py 

Seems like that would do it, so long as the `run_experiment()` function incorporates the additions/changes correctly, https://github.com/jsalfity/sensitivity_games/blob/main/sensitivity_games/experiments/point_mass_regulation_train.py#L46-L64. Important to note that `dynamics` and `controller` can't be recreated every epoch, because the torch variables are in `dynamics.theta` and `controller.K`.

It seems like things are getting messy. Wondering if there is a better way or time to refactor ?
Opened up a new branch, `sweep_initial_final_states_during_training`

### PR DESCRIPTION
**Pull request by jsalfity**
randomized `x0` within a range, I think this is giving the backprop algorithm more to work with. K is converging well, very close to DARE solutions and gradients are nearly zero. After 30k epochs ---

```sh
K value: tensor([[ 9.3169e-01,  1.4124e+00,  1.8969e-08, -5.5819e-08],
        [ 2.3500e-08, -1.5418e-07,  9.3169e-01,  1.4124e+00]],
       requires_grad=True)
K grad: tensor([[-6.3556e-07,  3.0429e-07, -7.6407e-08,  3.6584e-08],
        [ 2.9133e-07, -1.9655e-07,  3.5032e-08, -2.3634e-08]])

DARE K: tensor([[ 9.1704e-01,  1.6821e+00, -2.8130e-15, -1.4998e-15],
        [-1.0866e-15, -1.3271e-15,  9.1704e-01,  1.6821e+00]])
```

### CODE REVIEW

File: sensitivity_games/tests/test_point_mass.py [hunk 1]
```diff
@@ -97,9 +98,20 @@ def test_dm_curvature_should_be_positive(self):
 
     def test_point_mass_should_converge_to_dare_k(self):
         # TODO: push out T until close, find 'infinite' T
-        np.testing.assert_allclose(self.controller.K.detach(),
-                                   self.optimal_controller.K.detach(),
-                                   atol=zero_tolerance)
+
+        x1 = np.where(self.controller.K.detach() > zero_tolerance,
```
Comment (dfridovi) [inline sensitivity_games/tests/test_point_mass.py:104]: suggest adding a comment here to clarify

**Review by dfridovi — APPROVED** [review]
Simple, clean, I like it!

File: sensitivity_games/tests/test_point_mass.py (hunk 1 above)
Comment (jsalfity) [reply sensitivity_games/tests/test_point_mass.py:104]: yea, its a hack to drop the small zeros, i.e 1e-03, 1e-07, so that `np.testing.assert_allclose()` works properly. The cleaner way is probably somewhere in the torch docs and has to do with understanding torch tensors better, but this works for now, unless you think of something better.

Comment (dfridovi) [reply sensitivity_games/tests/test_point_mass.py:104]: makes sense - I guessed that was what was going on, but I still think a comment in the src is warranted
